### ISSUE TITLE
BPF ring buffer support

### ISSUE THREAD

**Issue by pwang7** [rust-bpf/rust-bcc#157]
One line summary of the issue here
BPF_RINGBUF_OUTPUT support

briefly answer these questions:

* what did you expect?
Does rust-bcc support BPF_RINGBUF_OUTPUT?

* what actually happened?
I have a BCC code using BPF_RINGBUF_OUTPUT, and I'd like to rewrite it using rust-bcc, but I didn't find BPF_RINGBUF_OUTPUT in rust-bcc.

* what steps can we take to reproduce the behavior you saw?
NONE

**Comment by javierhonduco** [comment]
`rust-bpf` does not have support for bpf ring buffers yet. Implementing this feature should be similar to how it was done in the BCC's python version in https://github.com/iovisor/bcc/commit/fe730f29f14bef8b5ffe1112c578df876c44d22d.

I think it would be a great feature to have. Being in parity with the Python wrapper will be great, and the performance benefits of BPF ring buffer would also be great!

**Comment by pwang7** [comment]
@javierhonduco Thanks for your reply!

Should I define [ringbuf related functions](https://github.com/iovisor/bcc/blob/fe730f29f14bef8b5ffe1112c578df876c44d22d/src/cc/libbpf.h#L124-L133) in bcc-sys first?

BTW, ringbuf is only available in BCC 0.16.0, right?

**Comment by brayniac** [comment]
> @javierhonduco Thanks for your reply!
> 
> Should I define [ringbuf related functions](https://github.com/iovisor/bcc/blob/fe730f29f14bef8b5ffe1112c578df876c44d22d/src/cc/libbpf.h#L124-L133) in bcc-sys first?
> 
> BTW, ringbuf is only available in BCC 0.16.0, right?

bcc-sys is generated bindings. All functions/constants should already be in there if they are in a supported bcc version. The only changes required should be in this repo.

The docs.rs generated documentation for bcc-sys will be for the latest version of bcc supported. If you don't see it there, open an issue in that repo and we can see if there's just a version we don't have generated bindings for yet.

I would honestly need to look at the main bcc documentation to learn more about the ring buffer, as my primary use has been in getting counters and histogram data from bpf programs. 

Let me know if you have additional questions. I'll respond as time allows. Thanks for your interest in the project!

**Comment by pwang7** [comment]
@brayniac Thanks Brian!

I'm interested in implementing ring buffer for rust-bpf.

I've opened an [issue](https://github.com/rust-bpf/bcc-sys/issues/55) in bcc-sys about missing ring buffer API.

BTW, if I implemented ring buffer for rust-bpf, how to run it using GitHub action? AFAIK, BPF ring buffer is available with kernel 5.8+. The latest Ubuntu 20 ships with kernel 5.4. I don't know whether GitHub action can upgrade kernel or not.

**Comment by javierhonduco** [comment]
> BTW, if I implemented ring buffer for rust-bpf, how to run it using GitHub action? AFAIK, BPF ring buffer is available with kernel 5.8+. The latest Ubuntu 20 ships with kernel 5.4. I don't know whether GitHub action can upgrade kernel or not.

Unfortunately this can't be done easily. The only way I can think of is running a VM inside of actions, making the testing process slower and more complex 😢.

**Comment by pwang7** [comment]
> > BTW, if I implemented ring buffer for rust-bpf, how to run it using GitHub action? AFAIK, BPF ring buffer is available with kernel 5.8+. The latest Ubuntu 20 ships with kernel 5.4. I don't know whether GitHub action can upgrade kernel or not.
> 
> Unfortunately this can't be done easily. The only way I can think of is running a VM inside of actions, making the testing process slower and more complex .

I tried installing KVM in github action, seems doesn't work. GitHub action environment is inside Microsoft Azure VM, and it doesn't support running VM in VM.

**Comment by brayniac** [comment]
Yeah - CI is going to be tough to handle on this. We should see what upstream bcc is doing. I'm a little nervous about having code paths we can't test in CI - esp ones that I'm unlikely to exercise regularly myself. That said, making this crate feature complete is something we have to balance against that. 

I'd be willing to add travisci back if that helps - wasn't able to find their kernel version list easily, but maybe with it running on GCP things would be easier?

**Comment by javierhonduco** [comment]
Upstream runs their own build bots with custom-ish distros (?) visa Jenkins CI https://buildbot.iovisor.org/jenkins/

**Comment by brayniac** [comment]
> Upstream runs their own build bots with custom-ish distros (?) visa Jenkins CI https://buildbot.iovisor.org/jenkins/

Bummer. Makes sense though. Maybe we hide the ring buffer support behind a feature flag? Open to other thoughts.

**Comment by pwang7** [comment]
> > Upstream runs their own build bots with custom-ish distros (?) visa Jenkins CI https://buildbot.iovisor.org/jenkins/
> 
> Bummer. Makes sense though.
> 
> EDIT: I guess we can just run without CI? Hiding behind a feature flag is likely only going to make maintenance and docs more complicated. CI will ensure we don't break anything we can test. Ideally we can do something in travis for this, but otherwise - let's just roll with it.

It seems Travis CI supports nested virtualization, but GitHub action does not, see the GitHub action issue [here](https://github.com/actions/virtual-environments/issues/183#issuecomment-567744066).

Therefore, we can try to install KVM inside Travis CI environment with latest kernel to test rust-bpf.

**Comment by brayniac** [comment]
> > > Upstream runs their own build bots with custom-ish distros (?) visa Jenkins CI https://buildbot.iovisor.org/jenkins/
> > 
> > 
> > Bummer. Makes sense though.
> > EDIT: I guess we can just run without CI? Hiding behind a feature flag is likely only going to make maintenance and docs more complicated. CI will ensure we don't break anything we can test. Ideally we can do something in travis for this, but otherwise - let's just roll with it.
> 
> It seems Travis CI supports nested virtualization, but GitHub action does not, see the GitHub action issue [here](https://github.com/actions/virtual-environments/issues/183#issuecomment-567744066).
> 
> Therefore, we can try to install KVM inside Travis CI environment with latest kernel to test rust-bpf.

Great! I'm open to re-adding some limited coverage in travis for newer kernel (and functionality dependent on new kernels). Hopefully I can make some time (or you can try) to regenerate the 0.16.0 bindings for bcc-sys. That should allow us to unblock the addition here. I'm unlikely to have time until next week.

**Comment by pwang7** [comment]
> > > > Upstream runs their own build bots with custom-ish distros (?) visa Jenkins CI https://buildbot.iovisor.org/jenkins/
> > > 
> > > 
> > > Bummer. Makes sense though.
> > > EDIT: I guess we can just run without CI? Hiding behind a feature flag is likely only going to make maintenance and docs more complicated. CI will ensure we don't break anything we can test. Ideally we can do something in travis for this, but otherwise - let's just roll with it.
> > 
> > 
> > It seems Travis CI supports nested virtualization, but GitHub action does not, see the GitHub action issue [here](https://github.com/actions/virtual-environments/issues/183#issuecomment-567744066).
> > Therefore, we can try to install KVM inside Travis CI environment with latest kernel to test rust-bpf.
> 
> Great! I'm open to re-adding some limited coverage in travis for newer kernel (and functionality dependent on new kernels). Hopefully I can make some time (or you can try) to regenerate the 0.16.0 bindings for bcc-sys. That should allow us to unblock the addition here. I'm unlikely to have time until next week.

Another possible solution to run BPF with new kernel is to use Katana container, it supports customize kernel version, see the link [here](https://github.com/kata-containers/packaging/tree/master/kernel). 

I'll have some time next week to try to implement ring buffer for rust-bpf. 

BTW, how to re-generate bcc-sys, using bindgen?

**Comment by brayniac** [comment]
There should be directions in the readme for the bcc-sys repo. Update the
issue in that repo if they're unclear, I'd like to make it easy for others.

**Comment by pwang7** [comment]
@brayniac I just find out that [BCC 0.17.0](https://github.com/iovisor/bcc/tree/v0.17.0) is released. Should I generate bcc-sys for 0.17.0 only or generate bcc-sys for both 0.17.0 and 0.16.0?

**Comment by pwang7** [comment]
I got some error when generating bcc-sys for BCC 0.17.0.
```
error: failed to run custom build command for `bcc-sys v0.16.2`

Caused by:
  process didn't exit successfully: `/tmp/rust_cargo/target/debug/build/bcc-sys-2cbfa644f6f81e43/build-script-build` (exit code: 101)
  --- stdout
  cargo:rustc-link-lib=bcc

  --- stderr
  thread 'main' panicked at 'Unable to get layout information?', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.55.1/src/codegen/mod.rs:1778:33
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I'll implement ring buffer support first, and fix this issue later.

**Comment by brayniac** [comment]
With #159 merged, the ring buffer support can now be implemented in this crate.

**Comment by pwang7** [comment]
> With #159 merged, the ring buffer support can now be implemented in this crate.

Thanks! I just submitted a [draft PR](https://github.com/rust-bpf/rust-bcc/pull/160), it implemented ring buffer for rust-bcc.
I'll try to add integration test using GitHub action or Travis CI next.

**Comment by pwang7** [comment]
@brayniac  I doubt [here](https://github.com/rust-bpf/rust-bcc/blob/5828af4c9b327ffc7f618c3892ce0cf19084f9b5/src/perf_event/mod.rs#L36) might have memory leak

**Comment by brayniac** [comment]
> @brayniac I doubt [here](https://github.com/rust-bpf/rust-bcc/blob/5828af4c9b327ffc7f618c3892ce0cf19084f9b5/src/perf_event/mod.rs#L36) might have memory leak

I'm not sure what you mean. Definitely doesn't seem relevant to this particular issue though?

Please try to keep communication split up a little better. It'll make it easier on me and anybody else trying to follow along.

I prefer separate issues, if you suspect there's a memory leak in this crate, open a new issue to track it.

Keep things separated by repository - for example, we don't need to know about the bcc-sys build issue for bcc 0.17.0 in this issue - but that should be tracked in the bcc-sys repository.

Thanks

**Comment by brayniac** [comment]
Fixed in #160
